File: 2 - Regression/8 - Decision Tree Regression/decision_tree_regression.R
# Imports
# For decision tree
# installed.packages('rpart')
library('rpart')
# For random forest
#install.packages('randomForest')
library(randomForest)

# Data loading
data = read.csv('C:/DOC/Workspace/Machine Learning A-Z Template Folder/Part 2 - Regression/Section 8 - Decision Tree Regression/Position_Salaries.csv')
data = data[-1]

# Decision tree
regressor_tree = rpart(formula = Salary ~ ., data = data, control = rpart.control(minsplit = 1))

# Random forest
regressor_forest = randomForest(x = data[-2], y = data$Salary, ntree = 500)

# Making predicitons
x_grid = data.frame(Level = seq(min(data$Level), max(data$Level), 0.001))
pred_tree = predict(regressor_tree, x_grid)
pred_forest = predict(regressor_forest, x_grid)

# Plot
ggplot() + 
  geom_point(aes(x = data$Level, y = data$Salary), color = 'red') +
  geom_line(aes(x = x_grid$Level, y = pred_tree), color = 'yellow') + 
  geom_line(aes(x = x_grid$Level, y = pred_forest), color = 'orange') + 
  ggtitle('Decision tree') + 
  xlab('Career salary') + 
  ylab('Salary')
  

File: 2 - Regression/8 - Decision Tree Regression/decision_tree_regression.py
# -*- coding: utf-8 -*-
"""
Created on Wed Aug 21 13:09:22 2019

@author: VLADRARESDANAILA
"""

import numpy as np
import pandas as pd
import sklearn as sk
import sklearn.tree
import sklearn.ensemble
import sklearn.preprocessing
import matplotlib.pyplot as plt

data = pd.read_csv('C:/DOC/Workspace/Machine Learning A-Z Template Folder/Part 2 - Regression/Section 8 - Decision Tree Regression/Position_Salaries.csv')
x, y = data.iloc[:, 1].values, data.iloc[:, -1].values
x, y = np.expand_dims(x, axis = 1), np.expand_dims(y, axis = 1)

# Decision tree
regressor_tree = sk.tree.DecisionTreeRegressor() 
regressor_tree.fit(x, y)

# Random forest
regressor_forest = sk.ensemble.RandomForestRegressor()
regressor_forest.fit(x, y)

# Prediction
x_grid = np.arange(x.min(), x.max(), step = 0.0001)
x_grid = x_grid.reshape((x_grid.shape[0], 1))
pred_tree = regressor_tree.predict(x_grid)
pred_forst = regressor_forest.predict(x_grid)

# Plot
plt.scatter(x, y, color = 'red')
plt.plot(x_grid, pred_tree, color = 'yellow')
plt.plot(x_grid, pred_forst, color = 'orange')
plt.title('Decision tree regression')
plt.xlabel('Career level')
plt.ylabel('salary')
plt.show()